<h1><span style="text-decoration: underline;">H&eacute;ritage et polymorphisme</span></h1>
<p style="text-align: justify;">Une application <em><strong>Python</strong> </em>est compos&eacute;e d&rsquo;un ensemble d&rsquo;objets. Un des int&eacute;r&ecirc;ts de la programmation orient&eacute;e objet (<em><strong>POO</strong></em>) r&eacute;side dans les relations que ces objets entretiennent les uns avec les autres. Ces relations sont construites par les d&eacute;veloppeurs et constituent ce que l&rsquo;on appelle l&rsquo;architecture d&rsquo;une application. Il existe deux relations fondamentales en <em><strong>POO</strong> </em>:</p>
<ul>
<li style="text-align: justify;"><span style="text-decoration: underline;"><em><strong>a un(e) (has-a)</strong></em></span></li>
</ul>
<blockquote>
<p style="text-align: justify;">Cette relation permet de cr&eacute;er une relation de d&eacute;pendance d&rsquo;une classe envers une autre. Une classe a besoin des services d&rsquo;une autre classe pour r&eacute;aliser sa fonction. On parle &eacute;galement de relation de composition pour d&eacute;signer ce type de relation.</p>
<p style="text-align: justify;">La relation &laquo;&nbsp;<em><strong>a un</strong></em>&nbsp;&raquo; implique simplement qu&rsquo;un objet poss&egrave;de un attribut qui r&eacute;f&eacute;rence un autre objet. Par exemple, si une programme d&eacute;clare une classe<span>&nbsp;</span><em><code class="docutils literal notranslate"><span class="pre">Personne</span></code></em>, cette classe peut avoir une propri&eacute;t&eacute; pour stocker l&rsquo;adresse sous la forme d&rsquo;une classe adresse:</p>
</blockquote>

In [2]:
class Personne:

    def __init__(self, nom, prenom, adresse):
        self.nom = nom
        self.prenom = prenom
        self.adresse = adresse


class Adresse:

    def __init__(self, rue, code_postal, ville):
        self.rue = rue
        self.code_postal = code_postal
        self.ville = ville


adresse = Adresse(rue="4 rue d'ici", code_postal="78000", ville="Paris")
personne = Personne(prenom="Jean", nom="Dumond", adresse=adresse)

<blockquote>
<p style="text-align: justify;">Le code ci-dessus illustre la relation de composition et on peut bien dire qu&rsquo;une <em><strong><code>personne</code>&nbsp;</strong></em>a une&nbsp;<code><em><strong>adresse</strong></em></code>.</p>
</blockquote>
<ul>
<li style="text-align: justify;"><span style="text-decoration: underline;"><em><strong>est un(e) (is-a)</strong></em></span></li>
</ul>
<blockquote>
<p style="text-align: justify;">Cette relation permet de cr&eacute;er une cha&icirc;ne de relation d&rsquo;identit&eacute; entre des classes. Elle indique qu&rsquo;une classe peut &ecirc;tre assimil&eacute;e &agrave; une autre classe qui correspond &agrave; une notion plus abstraite ou plus g&eacute;n&eacute;rale. On parle<span>&nbsp;</span><strong>d&rsquo;h&eacute;ritage</strong><span>&nbsp;</span>pour d&eacute;signer le m&eacute;canisme qui permet d&rsquo;impl&eacute;menter ce type de relation.&nbsp;</p>
</blockquote>
<div class="section" id="l-heritage">
<h2 style="text-align: justify;"><span style="text-decoration: underline;">1. L&rsquo;h&eacute;ritage</span><a class="headerlink" href="https://gayerie.dev/docs/python/python3/heritage.html#l-heritage" title="Lien permanent vers ce titre"></a></h2>
<p style="text-align: justify;">L&rsquo;h&eacute;ritage est donc le m&eacute;canisme qui permet de traduire une relation de type &laquo;&nbsp;<em><strong>est un(e)</strong></em>&nbsp;&raquo;.</p>
<p style="text-align: justify;">Prenons l&rsquo;exemple d&rsquo;une classe<span>&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Voiture</span></code></strong></em><span>&nbsp;</span>et d&rsquo;une classe<span>&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Vehicule</span></code></strong></em>. Il est possible de dire qu&rsquo;une <code><em><strong>voiture</strong></em></code><span>&nbsp;</span><strong>est un</strong><span>&nbsp;</span><code><em><strong>v&eacute;hicule</strong></em></code>. On peut traduire cette relation en <em><strong>Python</strong></em>. On dit que<span>&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Vehicule</span></code></strong></em><span>&nbsp;</span>est la classe parente de<span>&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Voiture</span></code></strong></em><span>&nbsp;</span>ou la g&eacute;n&eacute;ralisation. Sym&eacute;triquement, on dit que<span>&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Voiture</span></code></strong></em><span>&nbsp;</span>est la <em><strong>classe enfant (ou fille)</strong></em> de<span>&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Vehicule</span></code></strong></em><span>&nbsp;</span>ou la sp&eacute;cialisation. On dit aussi que la classe<span>&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Vehicule</span></code></strong></em><span>&nbsp;</span>est la<span>&nbsp;</span><strong>super classe</strong><span>&nbsp;</span>de<span>&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Voiture</span></code></strong></em>.</p>
</div>

In [4]:
class Vehicule:
    pass


class Voiture(Vehicule):
    pass

<p style="text-align: justify;">Cette relation se traduit en pr&eacute;cisant le nom de la classe parente entre parenth&egrave;ses apr&egrave;s le nom de la classe.</p>
<p style="text-align: justify;">Si nous cr&eacute;ons un objet &agrave; partir de la classe<span>&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Voiture</span></code></strong></em>, cet objet est &agrave; la fois de type<span>&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Vehicule</span></code></strong></em><span>&nbsp;</span>et de type<span>&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Voiture</span></code></strong></em>.</p>

In [5]:
v = Voiture()
print(isinstance(v, Voiture))
print(isinstance(v, Vehicule))

True
True


<p><span>Le mot de h&eacute;ritage vient du fait qu&rsquo;une classe fille h&eacute;rite des attributs et des comportements de sa classe m&egrave;re.</span></p>

In [7]:
class Vehicule:

    def __init__(self):
        self._vitesse = 0

    @property
    def vitesse(self):
        return self._vitesse

    def accelerer(self, delta_vitesse):
        self._vitesse += delta_vitesse

    def decelerer(self, delta_vitesse):
        self._vitesse -= delta_vitesse


class Voiture(Vehicule):
    pass

In [8]:
v = Voiture()
v.vitesse

0

In [10]:
v.accelerer(40)
v.vitesse

120

<p>Les objets de type<span>&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Voiture</span></code></strong></em><span>&nbsp;</span>disposent des m&ecirc;mes comportements, propri&eacute;t&eacute;s et attributs d&eacute;clar&eacute;s dans la classe<span>&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Vehicule</span></code></strong></em>.</p>
<p>La classe fille peut fournir ses propres comportements et ses propres attributs:</p>

In [7]:
class Voiture(Vehicule):

    def klaxonner(self):
        print("tût tût !")

<h3><span style="text-decoration: underline;">1.1. Constructeur et h&eacute;ritage</span></h3>
<p style="text-align: justify;"><span>Le constructeur est, comme toutes les m&eacute;thodes, h&eacute;rit&eacute;. Cependant que ce passe-t-il si nous souhaitons ajouter une constructeur dans notre classe&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Voiture</span></code></strong></em><span>?</span></p>

In [8]:
class Voiture(Vehicule):

    def __init__(self, klaxon="tût tût !"):
        self.klaxon = klaxon

    def klaxonner(self):
        print(self.klaxon)

In [9]:
v = Voiture()
v.klaxonner()

tût tût !


In [10]:
v.vitesse

AttributeError: 'Voiture' object has no attribute '_vitesse'

<p style="text-align: justify;"><span>Cela ne fonctionne plus car le constructeur d&eacute;clar&eacute; dans la classe&nbsp;</span><em><strong><code>Voiture</code></strong></em><span>&nbsp;remplace et cache celui h&eacute;rit&eacute; de la classe&nbsp;</span><em><strong><code>Vehicule</code></strong></em><span>&nbsp;et l&rsquo;initialisation de l&rsquo;attribut&nbsp;</span><em><strong><code>_vitesse</code></strong></em><span>&nbsp;n&rsquo;est plus faite. Quand une classe h&eacute;rite d&rsquo;une autre classe, elle a la responsabilit&eacute; de s&rsquo;assurer que le constructeur de la classe parente est appel&eacute; pour garantir l&rsquo;initialisation. Pour cela, <em><strong>Python 3</strong></em> fournit la fonction&nbsp;</span><em><strong><code>super()</code></strong></em><span>&nbsp;qui retourne le type de la super classe. L&rsquo;impl&eacute;mentation correcte de la classe&nbsp;</span><em><strong><code>Voiture</code></strong></em><span>&nbsp;doit donc &ecirc;tre:</span></p>

In [11]:
class Voiture(Vehicule):

    def __init__(self, klaxon="tût tût !"):
        super().__init__()
        self.klaxon = klaxon

    def klaxonner(self):
        print(self.klaxon)

In [12]:
v = Voiture()
v.klaxonner()

tût tût !


In [13]:
v.vitesse

0

<p style="text-align: justify;"><span>Il est donc facile de passer des param&egrave;tres &agrave; l&rsquo;initialisation de la classe parente si cette derni&egrave;re attend des param&egrave;tres de constructeur.</span></p>

In [14]:
class Vehicule:

    def __init__(self, marque=None, vitesse_initiale=0):
        self.marque = marque
        self._vitesse = vitesse_initiale

    @property
    def vitesse(self):
        return self._vitesse

    def accelerer(self, delta_vitesse):
        self._vitesse += delta_vitesse

    def decelerer(self, delta_vitesse):
        self._vitesse -= delta_vitesse


class Voiture(Vehicule):

    def __init__(self, marque=None, vitesse_initiale=0, klaxon="tût tût !"):
        super().__init__(marque, vitesse_initiale)
        self.klaxon = klaxon

    def klaxonner(self):
        print(self.klaxon)

In [15]:
v = Voiture("De Lorean", 88.0)
v.vitesse

88.0

In [16]:
v.marque

'De Lorean'

In [17]:
v.klaxonner()

tût tût !


<p style="text-align: justify;">Pour cha&icirc;ner l&rsquo;appel des constructeurs, il est &eacute;galement possibled&rsquo;appeler directement le constructeur en d&eacute;signant la classe parente:</p>

    
```python
class Voiture(Vehicule):
    def__init__(self,marque=None,vitesse_initiale=0,klaxon="tût tût !"):
        Vehicule.__init__(marque,vitesse_initiale)
        self.klaxon=klaxon
```
<p style="text-align: justify;">Cette &eacute;criture est cependant consid&eacute;r&eacute;e comme obsol&egrave;te en <em><strong>Python 3</strong></em> gr&acirc;ce &agrave; la fonction&nbsp;<em><strong><code>super()</code></strong></em>.</p>
</div>
</div>
<div>
<h3 style="text-align: justify;"><span style="text-decoration: underline;">1.2. H&eacute;ritage et mutualisation de code</span></h3>
<p style="text-align: justify;">Un des int&eacute;r&ecirc;ts de l&rsquo;h&eacute;ritage est de permettre la r&eacute;utilisation de code. Maintenant que nous avons d&eacute;fini un&nbsp;<em><strong><code>Vehicule</code></strong></em>, nous pouvons imaginer diverses classes appropri&eacute;es pour notre syst&egrave;me et qui sont des v&eacute;hicules.</p>
<p style="text-align: justify;"></p>
<p style="text-align: justify;"></p>
</div>



In [18]:
class Voiture(Vehicule):

    def __init__(self, marque=None, vitesse_initiale=0, klaxon="tût tût !"):
        Vehicule.__init__(marque, vitesse_initiale)
        self.klaxon = klaxon

In [19]:
class Charette(Vehicule):

    def __init__(self, vitesse_initiale=0):
        super().__init__(vitesse_initiale=vitesse_initiale)

In [20]:
c = Charette(1)
c.accelerer(5)
c.vitesse

6

<h3><span style="text-decoration: underline;">1.3. La classe&nbsp;object</span></h3>
<p style="text-align: justify;">Python d&eacute;finit la classe&nbsp;<em><strong><code><span>object</span></code></strong></em>. Toutes les classes h&eacute;rite directement ou indirectement de cette classe. Si une classe ne d&eacute;clare aucune classe parente alors sa classe parente est&nbsp;<span><em><strong><code><span>object</span></code></strong></em></span>.</p>

In [21]:
class MaClasse:
    pass

<p>La d&eacute;claration ci-dessus est strictement &eacute;quivalente &agrave; :</p>

In [22]:
class MaClasse(object):
    pass

<p style="text-align: justify;"><span>Il est possible de cr&eacute;er des instances de la classe&nbsp;</span><code><em><strong>object</strong></em></code><span>. Cela reste d&rsquo;un usage limit&eacute; car cette classe n&rsquo;offre aucune m&eacute;thode particuli&egrave;re et il n&rsquo;est pas possible d&rsquo;ajouter dynamiquement des attributs &agrave; ses instances.</span></p>

In [23]:
o = object()
o.nom = "Un nom"

AttributeError: 'object' object has no attribute 'nom'

<h2><span style="text-decoration: underline;">2. Le polymorphisme</span></h2>
<p style="text-align: justify;">Le polymorphisme est un m&eacute;canisme important dans la programmation objet. Il permet de modifier le comportement d&rsquo;une classe fille par rapport &agrave; sa classe m&egrave;re. Le polymorphisme permet d&rsquo;utiliser l&rsquo;h&eacute;ritage comme un m&eacute;canisme d&rsquo;extension en adaptant le comportement des objets.</p>
<p style="text-align: justify;">Prenons l&rsquo;exemple de la classe<span>&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Animal</span></code></strong></em>. Cette classe offre une m&eacute;thode<span>&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">crier</span></code></strong></em>. Pour simplifier notre exemple, la m&eacute;thode se contente d&rsquo;&eacute;crire le cri de l&rsquo;animal sur la sortie standard.</p>

In [24]:
class Animal:

    def crier(self):
        print("un cri d'animal")

<p style="text-align: justify;"><span>Nous pouvons cr&eacute;er les classes&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Chien</span></code></strong></em><span>&nbsp;et&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Chat</span></code></strong></em><span>&nbsp;qui h&eacute;ritent toutes deux de la classe&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Animal</span></code></strong></em><span>. Ces classes peuvent &ecirc;tre des sp&eacute;cialisations de la classe&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Animal</span></code></strong></em><span>&nbsp;en ce qu&rsquo;elles peuvent&nbsp;</span><strong>red&eacute;finir</strong><span>&nbsp;(</span><strong><em>override</em></strong><span>) le comportement de la m&eacute;thode&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">crier</span><span class="pre">()</span></code></strong></em><span>.</span></p>

In [25]:
class Chien(Animal):

    def crier(self):
        print("whouaf whouaf !")


class Chat(Animal):

    def crier(self):
        print("miaou !")

In [26]:
a = Animal()
animal = Animal()
animal.crier()

un cri d'animal


In [27]:
animal = Chien()
animal.crier()

whouaf whouaf !


In [28]:
animal = Chat()
animal.crier()

miaou !


<p style="text-align: justify;">Un <code><em><strong>chat</strong></em></code> est bien un <code><em><strong>animal</strong></em></code>, un <code><em><strong>chien</strong></em></code> est bien un <code><em><strong>animal</strong></em></code>. Les objets de ces types disposent bien de la m&eacute;thode<span>&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">crier()</span></code></strong></em><span>&nbsp;</span>mais son comportement est polymorphe. Il d&eacute;pend du type r&eacute;el de l&rsquo;objet.</p>
<p style="text-align: justify;">&Agrave; noter qu&rsquo;il est toujours possible d&rsquo;appeler la m&eacute;thode parente dans la classe enfant gr&acirc;ce &agrave; la fonction<span>&nbsp;</span><em><strong><code class="xref py py-func docutils literal notranslate"><span class="pre">super()</span></code></strong></em>.</p>

In [29]:
class Chien(Animal):

    def crier(self):
        print("whouaf whouaf !")

    def crier_comme_un_animal(self):
        super().crier()

In [30]:
chien = Chien()
chien.crier()

whouaf whouaf !


In [31]:
chien.crier_comme_un_animal()

un cri d'animal


<div>
<h3 style="text-align: justify;"><span style="text-decoration: underline;">2.1 Polymorphisme et&nbsp;<em>duck typing</em></span></h3>
<p style="text-align: justify;">Pour les langages de programmation &agrave; typage fort comme le <em><strong>C++</strong></em> ou le <em><strong>Java</strong></em>, le polymorphisme a des implications tr&egrave;s fortes et il permet d&rsquo;introduire une grande souplesse dans les relations entre types. Pour les langages de programmation &agrave; typage dynamique comme <em><strong>Python</strong></em>, le polymorphisme, m&ecirc;me s&rsquo;il reste une notion importante, a moins d&rsquo;impact.</p>
<p style="text-align: justify;">En <em><strong>Python</strong></em>, on pr&eacute;f&egrave;re souvent le principe du <em><strong>duck typing(ou typage canard)</strong></em>. Cette approche repose sur le principe suivant :</p>
<blockquote>
<p>Si je vois un oiseau qui vole comme un canard, cancane comme un canard, et nage comme un canard, alors j&rsquo;appelle cet oiseau un canard.</p>
</blockquote>
<p style="text-align: justify;">Cela signifie que le type r&eacute;el est moins important que le comportement attendu. Pour reprendre notre exemple des v&eacute;hicules: si une classe offre des m&eacute;thodes pour acc&eacute;l&eacute;rer, d&eacute;c&eacute;l&eacute;rer et des propri&eacute;t&eacute;s pour la vitesse et la marque, alors c&rsquo;est qu&rsquo;il s&rsquo;agit d&rsquo;un v&eacute;hicule et peu importe que cette classe h&eacute;rite ou non de la classe&nbsp;<em><strong><code>Vehicule</code></strong></em>. Tous les objets de cette classe pourront &ecirc;tre utilis&eacute;s dans un contexte qui manipule des v&eacute;hicules.</p>
</div>
<div>
<h3 style="text-align: justify;"><span style="text-decoration: underline;">2.2. Masquer les attributs et les m&eacute;thodes pour les classe filles</span></h3>
<p style="text-align: justify;">Parfois, on ne souhaite pas qu&rsquo;une m&eacute;thode puisse &ecirc;tre red&eacute;finie ou qu&rsquo;un attribut puisse &ecirc;tre modifi&eacute; dans une classe fille. Pour cela, il suffit que le nom de la m&eacute;thode ou de l&rsquo;attribut commence par deux <em><strong>caract&egrave;res soulign&eacute;s (underscores)</strong></em>. Nous avons vu pr&eacute;c&eacute;demment que <em><strong>Python</strong></em>n&rsquo;a pas de m&eacute;canisme pour contr&ocirc;ler la visibilit&eacute; des &eacute;l&eacute;ments d&rsquo;une classe. Par convention, les d&eacute;veloppeurs signalent par un <em><strong>caract&egrave;re soulign&eacute; (underscore)</strong></em> le statut priv&eacute; d&rsquo;un attribut ou d&rsquo;une m&eacute;thode. Par contre le recours &agrave; deux caract&egrave;res soulign&eacute;s &agrave; un impact sur l&rsquo;interpr&eacute;teur. Ce dernier renomme la m&eacute;thode ou l&rsquo;attribut de la forme:</p>
</div>

```python
_<nom de la classe>__<nom>
```   

<p style="text-align: justify;">Ainsi si nous modifions le nom de la la classe&nbsp;<em><strong><code>Vehicule</code></strong></em>:</p>

In [32]:
class Vehicule:

    def __init__(self, marque=None, vitesse_initiale=0):
        self.marque = marque
        self.__vitesse = vitesse_initiale

    @property
    def vitesse(self):
        return self.__vitesse

    def accelerer(self, delta_vitesse):
        self.__vitesse += delta_vitesse

    def decelerer(self, delta_vitesse):
        self.__vitesse -= delta_vitesse

<p style="text-align: justify;"><span>L&rsquo;attribut&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">__vitesse</span></code></strong></em><span>&nbsp;sera renomm&eacute;&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">_Vehicule__vitesse</span></code></strong></em><span>. Donc, si un attribut&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">__vitesse</span></code></strong></em><span>&nbsp;est &eacute;galement utilis&eacute; dans une classe fille, l&rsquo;interpr&eacute;teur consid&eacute;rera qu&rsquo;il s&rsquo;agit d&rsquo;un attribut diff&eacute;rent.</span></p>
<h3 style="text-align: justify;"><span style="text-decoration: underline;">2.3. Les exceptions comme objets</span></h3>
<p style="text-align: justify;">En <em><strong>Python</strong></em>, une exception est un objet qui est directement ou indirectement une instance de la classe&nbsp;<em><strong><code>BaseException</code></strong></em>.</p>
<p style="text-align: justify;">Il existe beaucoup de classes pour repr&eacute;senter des exceptions en <em><strong>Python</strong></em>. Vous pouvez consulter la documentation pour conna&icirc;tre la hi&eacute;rarchie des exceptions. N&eacute;anmoins, il est tr&egrave;s simple de cr&eacute;er ses propres exceptions. Il est recommand&eacute; de cr&eacute;er des exceptions en h&eacute;ritant de <em><strong><code>Exception</code></strong></em> ou d&rsquo;une classe h&eacute;ritant de <em><strong><code>Exception</code></strong></em>.&nbsp;<em><strong><code>Exception</code></strong></em> est une classe qui h&eacute;rite de <em><strong><code>BaseException</code></strong></em>. Pour simplifier l&rsquo;impl&eacute;mentation, <em><strong><code>BaseException</code></strong></em> d&eacute;finit l&rsquo;attribut <em><strong><code>args</code></strong></em> qui contient tous les param&egrave;tres pass&eacute;s au constructeur.</p>

In [33]:
class MonException(Exception):
    pass

try:
    raise MonException("Mon message")
except MonException as e:
    print(e.args)
    # Affiche ('Mon message',)
    print(e)
    # Affiche Mon message

('Mon message',)
Mon message


<h2><span style="text-decoration: underline;">3. H&eacute;ritage multiple et&nbsp;<em>mixin</em></span></h2>
<p style="text-align: justify;">Une classe peut h&eacute;riter de plusieurs classes. Dans ce cas, elle h&eacute;ritera des m&eacute;thodes et des attributs de l&rsquo;ensemble de ces classes et de leurs classes parentes. Il suffit d&rsquo;indiquer la liste des classes parentes en les s&eacute;parant par une virgule.</p>

In [34]:
class Animal:

    def crier(self):
        pass


class Carnivore:

    def chasser(self):
        pass


class Chien(Animal, Carnivore):
    """Un chien qui est à la fois un animal et un carnivore"""

In [35]:
c = Chien()
c.chasser()
c.crier()

<p style="text-align: justify;"><span>La classe&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Chien</span></code></strong></em><span>&nbsp;h&eacute;rite des comportements mais aussi des attributs des classes parentes. La variable&nbsp;</span><code class="docutils literal notranslate"><span class="pre"><em>c</em></span></code><span>&nbsp;d&eacute;signe un objet qui est &agrave; la fois un&nbsp;</span><strong><code class="docutils literal notranslate"><span class="pre">Chien</span></code></strong><span>, un&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Animal</span></code></strong></em><span><em><strong>&nbsp;</strong></em>et un&nbsp;</span><em><strong><code class="docutils literal notranslate"><span class="pre">Carnivore</span></code></strong></em><span>.</span></p>

In [36]:
print(isinstance(c, Chien))
print(isinstance(c, Animal))
print(isinstance(c, Carnivore))

True
True
True


<h3><span style="text-decoration: underline;">L&rsquo;h&eacute;ritage en diamant</span></h3>
<p style="text-align: justify;">Il peut arriver qu&rsquo;une classe h&eacute;rite indirectement plusieurs fois de la m&ecirc;me classe. Reprenons notre exemple pr&eacute;c&eacute;dent et supposons que les classes&nbsp;<em><strong><code>Animal</code></strong></em> et&nbsp;<em><strong><code>Carnivore</code></strong></em> h&eacute;ritent toutes deux de la classe&nbsp;<em><strong><code>EtreVivant</code></strong></em> et que la classe&nbsp;<code><span><em><strong>EtreVivant</strong></em></span></code> poss&egrave;de l&rsquo;attribut <code><span><em><strong>point_de_vie</strong></em></span></code>.</p>

In [37]:
class EtreVivant:

    def __init__(self):
        self.point_de_vie = 100

class Animal(EtreVivant):

    def dormir(self):
        self.point_de_vie += 1

class Carnivore(EtreVivant):

    def chasser(self):
        self.point_de_vie -= 1


class Chien(Animal, Carnivore):
    """Un chien qui est à la fois un animal et un carnivore"""

<p>Dans ce cas, la classe&nbsp;<code><span><em><strong>Chien</strong></em></span></code> h&eacute;rite deux fois de la classe&nbsp;<em><code>EtreVivant</code></em>. &Agrave; cause de la repr&eacute;sentation graphique d&rsquo;une telle situation, on appelle ce cas particulier&nbsp;<strong>l&rsquo;h&eacute;ritage en diamant</strong>.</p>
<p><em><strong>Python</strong></em> r&eacute;sout cette situation en consid&eacute;rant qu&rsquo;une classe ne peut pas directement ou indirectement h&eacute;riter plusieurs fois d&rsquo;une m&ecirc;me classe. Donc, les instances de la classe&nbsp;<code><span><em><strong>Chien</strong></em></span></code> ne poss&eacute;deront qu&rsquo;un seul attribut&nbsp;<em><strong><code>point_de_vie</code></strong></em>.</p>

In [38]:
c = Chien()
c.chasser()
c.point_de_vie

99

In [39]:
c.dormir()
c.point_de_vie

100

<p style="text-align: justify;"><span>Mais que se passe-t-il si l&rsquo;h&eacute;ritage en diamant implique des m&eacute;thodes ? Par exemple, imaginons que la classe&nbsp;</span><code><span><em><strong>EtreVivant</strong></em></span></code><span>&nbsp;poss&egrave;de la m&eacute;thode&nbsp;</span><em><strong><code>se_nourrir()</code></strong></em><span>&nbsp;qui est red&eacute;finie &agrave; la fois dans la classe&nbsp;</span><code><span><em><strong>Animal</strong></em></span></code><span>&nbsp;et dans la classe&nbsp;</span><em><strong><code>Carnivore</code></strong></em><span>.</span></p>

In [40]:
class EtreVivant:

    def __init__(self):
        self.point_de_vie = 100

    def se_nourrir(self):
        self.point_de_vie += 1


class Animal(EtreVivant):

    def dormir(self):
        self.point_de_vie += 1

    def se_nourrir(self):
        self.point_de_vie += 5


class Carnivore(EtreVivant):

    def chasser(self):
        self.point_de_vie -= 1

    def se_nourrir(self):
        self.point_de_vie += 10


class Chien(Animal, Carnivore):
    """Un chien qui est à la fois un animal et un carnivore"""

<p>Que se passe-t-il si nous appelons la m&eacute;thode&nbsp;<em><strong><code>se_nourrir()</code>&nbsp;</strong></em>depuis une instance de&nbsp;<em><strong><code>Chien</code></strong></em>?</p>

In [41]:
c = Chien()
c.se_nourrir()
c.point_de_vie

105

<p style="text-align: justify;"><span>C&rsquo;est la m&eacute;thode de la classe&nbsp;</span><em><strong><code>Animal</code></strong></em><span>&nbsp;qui est ex&eacute;cut&eacute;e, passant la valeur de&nbsp;</span><em><strong><code>point_de_vie</code></strong></em><span>&nbsp;&agrave; <em><strong>105</strong></em>. Pour choisir quelle m&eacute;thode est appel&eacute;e, l&rsquo;interpr&eacute;teur se base sur un algorithme appel&eacute;&nbsp;</span><strong>method resolution order</strong><span>&nbsp;ou&nbsp;</span><strong>mro</strong><span>. Il s&rsquo;agit de chercher dans la liste de l&rsquo;h&eacute;ritage des classes la premi&egrave;re d&eacute;claration de la m&eacute;thode appel&eacute;e. On peut facilement conna&icirc;tre le&nbsp;</span><em>mro</em><span>&nbsp;d&rsquo;une classe car il existe la m&eacute;thode&nbsp;</span><em><strong><code>class.mro()</code></strong></em><span>&nbsp;qui retourne pr&eacute;cis&eacute;ment cette liste. Pour acc&eacute;der &agrave; la classe d&rsquo;une instance, on peut utiliser la fonction&nbsp;</span><em><strong><code>type</code></strong></em><span>.</span></p>

In [42]:
type(c).mro()

[__main__.Chien,
 __main__.Animal,
 __main__.Carnivore,
 __main__.EtreVivant,
 object]

<p>On voit que pour le type&nbsp;<em><strong><code>Chien</code></strong></em>, un appel &agrave; une m&eacute;thode conduit &agrave; chercher cette m&eacute;thode d&rsquo;abord dans la classe<code><span>Chien</span></code>et ensuite dans les classes&nbsp;<em><strong><code>Animal</code></strong></em>,&nbsp;<em><strong><code>Carnivore</code></strong></em>,&nbsp;<em><strong><code>EtreVivant</code></strong></em> et enfin&nbsp;<em><strong><code>object</code></strong></em>.</p>
<div>
<p>Si nous intervertissons l&rsquo;ordre d&rsquo;h&eacute;ritage de la d&eacute;claration de la classe&nbsp;<em><strong><code>Chien</code></strong></em>.</p>
</div>

In [43]:
class EtreVivant:

    def __init__(self):
        self.point_de_vie = 100

    def se_nourrir(self):
        self.point_de_vie += 1


class Animal(EtreVivant):

    def dormir(self):
        self.point_de_vie += 1

    def se_nourrir(self):
        self.point_de_vie += 5


class Carnivore(EtreVivant):

    def chasser(self):
        self.point_de_vie -= 1

    def se_nourrir(self):
        self.point_de_vie += 10


class Chien(Carnivore, Animal):
    """Un chien qui est à la fois un animal et un carnivore"""

In [44]:
c = Chien()
c.se_nourrir()
c.point_de_vie

110

In [45]:
type(c).mro()

[__main__.Chien,
 __main__.Carnivore,
 __main__.Animal,
 __main__.EtreVivant,
 object]

<div>
<p style="text-align: justify;">Maintenant, c&rsquo;est la m&eacute;thode&nbsp;<em><strong><code>se_nourrir()</code>&nbsp;</strong></em>de la classe <code><span><em><strong>Carnivore</strong></em></span></code> qui est appel&eacute;e pour un objet de type <em><strong><code>Chien</code></strong></em>. L&rsquo;ordre de d&eacute;claration de l&rsquo;h&eacute;ritage multiple est donc primordial !</p>
</div>
<div>
<h3 style="text-align: justify;"><span style="text-decoration: underline;">Appel des constructeurs</span></h3>
<p style="text-align: justify;">Si nous voulons d&eacute;clarer des constructeurs, il faut imp&eacute;rativement&nbsp;<strong>appeler le constructeur de la super classe</strong>.</p>
</div>

In [46]:
class EtreVivant:

    def __init__(self, point_de_vie):
        self.point_de_vie = point_de_vie

    def se_nourrir(self):
        self.point_de_vie += 1


class Animal(EtreVivant):

    def __init__(self, nom, point_de_vie):
        super().__init__(point_de_vie)
        self.nom = nom

    def dormir(self):
        self.point_de_vie += 1

    def se_nourrir(self):
        self.point_de_vie += 5


class Carnivore(EtreVivant):

    def chasser(self):
        self.point_de_vie -= 1

    def se_nourrir(self):
        self.point_de_vie += 10


class Chien(Carnivore, Animal):
    """Un chien qui est à la fois un animal et un carnivore"""

    def __init__(self, point_de_vie, nom):
        super().__init__(point_de_vie, nom)

In [47]:
c = Chien("Médor", 60)
c.nom

'Médor'

In [48]:
c.point_de_vie

60

In [49]:
c.se_nourrir()
c.point_de_vie

70

<div>
<p style="text-align: justify;">La fonction&nbsp;<code><span><em><strong>super()</strong></em></span></code> retourne le type suivant dans la liste du<em>mro</em>. La classe<code><span>Carnivore</span></code>ne proposant pas de constructeur, elle se contente de passer l&rsquo;appel au constructeur de la classe&nbsp;<code><span><em><strong>Animal</strong></em></span></code> qui est la classe suivante dans la liste&nbsp;<code><strong><em>mro</em></strong></code>.</p>
</div>
<div>
<h3 style="text-align: justify;"><span style="text-decoration: underline;">Les mixins</span></h3>
<p style="text-align: justify;">Un&nbsp;<em><code><strong>mixin</strong></code>&nbsp;</em>est une classe qui permet d&rsquo;ajouter des fonctionnalit&eacute;s suppl&eacute;mentaires. Il s&rsquo;agit simplement d&rsquo;une classe comme une autre mais qui n&rsquo;est pas destin&eacute;e &agrave; &ecirc;tre utilis&eacute;e directement pour cr&eacute;er des instances.</p>
<p style="text-align: justify;">Imaginons que nous d&eacute;sirions ajouter &agrave; certaines de nos classes la possibilit&eacute; d&rsquo;afficher leur &eacute;tat (la valeur des attributs) gr&acirc;ce &agrave; une m&eacute;thode&nbsp;<em><strong><code>afficher()</code></strong></em>. Nous pouvons cr&eacute;er un&nbsp;<em><code><strong>mixin</strong></code>&nbsp;</em>proposant cette m&eacute;thode. Ce&nbsp;<em><code><strong>mixin</strong></code>&nbsp;</em>accepte en param&egrave;tre le caract&egrave;re &agrave; utiliser comme s&eacute;parateur ainsi que l&rsquo;indentation &agrave; utiliser lors de l&rsquo;affichage.</p>
</div>

In [50]:
class Affichable:

    def __init__(self, *args, indentation=0, separateur="-", **kwargs):
        self.__indentation = indentation
        self.__separateur = separateur
        super().__init__(*args, **kwargs)

    def afficher(self):
        cls = type(self)
        separateur = self.__separateur * 80
        print(separateur)
        print("Objet de la classe", cls.__name__)
        print(cls.__doc__)
        print("mro : ", cls.mro())
        print()
        indentation = ' ' * self.__indentation
        for attr in self.__dict__:
            if not attr.startswith('_'):
                valeur = getattr(self, attr)
                print(f"{indentation}{attr} = {valeur}")
        print(separateur)

<p>Nous pouvons modifier notre impl&eacute;mentation de la classe&nbsp;<code><span><em><strong>Chien</strong></em></span></code> pour la rendre affichable :</p>

In [51]:
class Chien(Affichable, Animal):
    """Un chien qui est à la fois un animal et un carnivore"""

    def __init__(self, point_de_vie, nom, **kwargs):
        super().__init__(point_de_vie, nom, **kwargs)

<p>Ainsi nous pouvons afficher l&rsquo;&eacute;tat interne d&rsquo;un objet de type&nbsp;<em><strong><code><span>Chien</span></code></strong></em>.</p>

In [52]:
c = Chien(40, "Médor", indentation=4)
c.afficher()

--------------------------------------------------------------------------------
Objet de la classe Chien
Un chien qui est à la fois un animal et un carnivore
mro :  [<class '__main__.Chien'>, <class '__main__.Affichable'>, <class '__main__.Animal'>, <class '__main__.EtreVivant'>, <class 'object'>]

    point_de_vie = Médor
    nom = 40
--------------------------------------------------------------------------------


<p style="text-align: justify;">Il y a plusieurs points importants &agrave; remarquer dans l&rsquo;impl&eacute;mentation ci-dessus :</p>
<ol style="text-align: justify;">
<li>La classe<code><span>Chien</span></code>h&eacute;rite en premier du&nbsp;<em><code><strong>mixin</strong></code>&nbsp;<strong><code>Affichable</code></strong></em>. &Agrave; cause de ce que nous avons vu &agrave; propos du&nbsp;<code><strong><em>mro</em></strong></code>, un <em><code><strong>mixin</strong></code>&nbsp;</em>devrait toujours &ecirc;tre d&eacute;clar&eacute; avant la classe parente.</li>
<li>La classe&nbsp;<code><span><em><strong>Chien</strong></em></span></code> accepte comme dernier param&egrave;tre de constructeur le param&egrave;tre de compactage&nbsp;<em><strong><code>**kwargs</code></strong></em>. La classe&nbsp;<code><span><em><strong>Chien</strong></em></span></code> peut ainsi accepter les param&egrave;tres nomm&eacute;s&nbsp;<code><span><em><strong>identation</strong></em></span></code> et&nbsp;<code><span><em><strong>separateur</strong></em></span></code> sans avoir besoin de les r&eacute;p&eacute;ter.</li>
<li>La classe&nbsp;<code><span><em><strong>Affichage</strong></em></span></code> a un constructeur assez compliqu&eacute;.</li>
</ol>

```python
def __init__(self, *args, indentation=0, separateur="-", **kwargs):
    self.__indentation = indentation
    self.__separateur = separateur
    super().__init__(*args, **kwargs)
```

<p style="text-align: justify;">Les param&egrave;tres&nbsp;<code><span><em><strong>indentation</strong></em></span></code> et&nbsp;<em><strong><code>separateur</code></strong></em> sont plac&eacute;s apr&egrave;s l&rsquo;op&eacute;rateur de compactage <em><strong><code>*args</code></strong></em> ce qui implique qu&rsquo;il s&rsquo;agit de param&egrave;tres nomm&eacute;s. Le constructeur utilise &agrave; la fin la&nbsp; fonction&nbsp;<em><strong><code>super()</code></strong></em> pour appeler le constructeur suivant en lui passant tous les param&egrave;tres que le <em><code><strong>mixin</strong></code></em><em>&nbsp;</em>ne reconna&icirc;t pas. On voit qu&rsquo;il est tout &agrave; fait possible de concevoir un&nbsp;<em><code><strong>mixin</strong></code></em><i> </i>qui accepte des param&egrave;tres de constructeur tout en s&rsquo;int&eacute;grant &eacute;l&eacute;gamment &agrave; un h&eacute;ritage existant&hellip; au prix d&rsquo;une petite complexit&eacute; dans la d&eacute;claration du constructeur.</p>

<h3><span style="text-decoration: underline;">La m&eacute;ta-classe</span></h3>
<p>La m&eacute;ta-classe est un concept avanc&eacute; en <em><strong>Python&nbsp;</strong></em>qui n&rsquo;est que tr&egrave;s tr&egrave;s rarement utilis&eacute; directement par les d&eacute;veloppeurs.</p>
<p>En <em><strong>Python</strong></em>, les classes sont elles-m&ecirc;mes des objets qui h&eacute;ritent de<span>type</span>. Il est possible de sp&eacute;cifier le type dont doit h&eacute;riter l&rsquo;objet qui repr&eacute;sente la classe. On parle alors de&nbsp;<em><strong>m&eacute;ta-classe</strong></em>.</p>
<p>Une <em><strong>m&eacute;ta-classe</strong></em> est une classe qui d&eacute;crit une classe. Cela signifie que tous les attributs et toutes les m&eacute;thodes d&rsquo;une m&eacute;ta-classe seront les attributs et les m&eacute;thodes de la classe.</p>
<p>L&rsquo;usage de la&nbsp;<strong><em>m&eacute;ta-classe</em>&nbsp;</strong>permet de r&eacute;aliser des impl&eacute;mentations qui ne sont pas possibles avec une simple classe. Par exemple, il n&rsquo;est pas possible en <em><strong>Python</strong></em> de d&eacute;clarer une propri&eacute;t&eacute; de classe avec le d&eacute;corateur&nbsp;<em><strong><code>@property</code></strong></em>. Mais cela devient possible par le truchement d&rsquo;une <strong>m&eacute;ta-classe</strong>.</p>

In [54]:
class MetaclasseCompteur(type):
    """Une méta-classe pour aider à compter les instances créées."""

    def __init__(cls, *args, **kwargs):
        super().__init__(*args, **kwargs)
        cls._nb_instances = 0

    @property
    def nb_instances(cls):
        return cls._nb_instances

    def plus_une_instance(cls):
        cls._nb_instances += 1


class MaClasse(metaclass=MetaclasseCompteur):

    def __init__(self):
        MaClasse.plus_une_instance()

<p style="text-align: justify;"><span>Une <em><strong>m&eacute;ta-classe</strong></em> doit h&eacute;riter directement ou indirectement de&nbsp;</span><em><strong><code>type</code></strong></em><span>. &Agrave; la ligne 16, on d&eacute;clare la m&eacute;ta-classe de la classe&nbsp;</span><em><strong><code>MaClasse</code></strong></em><span>. Cette <em><strong>m&eacute;ta-classe</strong></em> va fournir la propri&eacute;t&eacute; de classe&nbsp;</span><em><strong><code>nb_instances</code></strong></em><span>&nbsp;ainsi que la m&eacute;thode de classe&nbsp;</span><em><strong><code>plus_une_instance()</code></strong></em><span>.</span></p>

In [55]:
MaClasse.nb_instances

0

In [56]:
m1 = MaClasse()
m2 = MaClasse()
m3 = MaClasse()
MaClasse.nb_instances

3

<h3 style="text-align: justify;"><span style="text-decoration: underline;">Classes et m&eacute;thodes abstraites</span></h3>
<p style="text-align: justify;">Comme <em><strong>Python&nbsp;</strong></em>est un langage &agrave; typage dynamique, nous avons d&eacute;j&agrave; dit plus haut que la plupart des d&eacute;veloppeurs favorisent le <code><em><strong>duck typing</strong></em></code>: le type r&eacute;el d&rsquo;un objet importe moins que le fait qu&rsquo;il produise les comportements attendus (c&rsquo;est-&agrave;-dire les m&eacute;thodes et les propri&eacute;t&eacute;s attendues). Cette approche am&egrave;ne ind&eacute;niablement plus de souplesse dans la conception des applications. Mais elle rend plus difficile la validation du code et donc cela peut aboutir &agrave; la production d&rsquo;un code moins robuste.</p>
<p style="text-align: justify;">Les langages de programmation &agrave; typage fort comme <em><strong>C++, Java ou C#</strong></em> introduisent tous le principe de classes abstraites et/ou d&rsquo;interfaces. Ce type d&rsquo;approche insiste sur le fait de pouvoir d&eacute;finir un type particulier contenant un certain nombre de m&eacute;thodes mais pour lesquelles on ne fournit aucune impl&eacute;mentation. Ces classes abstraites et ces interfaces sont ensuite h&eacute;rit&eacute;es ou impl&eacute;ment&eacute;es par d&rsquo;autres classes qui doivent fournir les impl&eacute;mentations des m&eacute;thodes attendues. Cela permet de garantir qu&rsquo;un objet aura bien les comportements attendus, c&rsquo;est-&agrave;-dire impl&eacute;mentera les m&eacute;thodes attendues. On parle parfois de programmation par contrat, dans le sens o&ugrave; ces classes abstraites et ces interfaces sont comme des contrats qui lient les objets qui les impl&eacute;mentent et les objets qui appellent ces m&eacute;thodes.</p>
<p style="text-align: justify;">En <em><strong>Python</strong></em>, le module&nbsp;<em><strong><code>abc</code></strong></em> permet de simuler ce type d&rsquo;approche. Le nom de ce module est la contraction de&nbsp;<strong><em>abstract base classes</em></strong>. Ce module fournit une <em><strong>m&eacute;ta-classe</strong></em> appel&eacute;e&nbsp;<code><span><em><strong>ABCMeta</strong></em></span></code> qui permet de transformer une classe <em><strong>Python</strong> </em>en classe abstraite. Ce module fournit &eacute;galement le d&eacute;corateur <em><strong><code>@abstractmethod</code></strong></em> qui permet de d&eacute;clarer comme abstraite une m&eacute;thode, une m&eacute;thode statique, une m&eacute;thode de classe ou une propri&eacute;t&eacute;. Cela signifie qu&rsquo;il n&rsquo;est pas possible de cr&eacute;er une instance d&rsquo;une classe qui h&eacute;rite d&rsquo;une classe abstraite tant que toutes les m&eacute;thodes abstraites ne sont pas impl&eacute;ment&eacute;es.</p>
<p style="text-align: justify;">Si nous reprenons notre exemple de la classe&nbsp;<em><strong><code>Animal</code></strong></em>. Cette classe d&eacute;clare une m&eacute;thode <em><strong><code>crier()</code>&nbsp;</strong></em>pour laquelle il n&rsquo;est pas vraiment possible de fournir une impl&eacute;mentation correcte pour un animal. On peut donc consid&eacute;rer que la classe&nbsp;<em><strong><code>Animal</code></strong></em> est abstraite en d&eacute;clarant que sa <em><strong>m&eacute;ta-classe</strong></em> est <em><strong><code>ABCMeta</code></strong></em> et d&eacute;clarer la <em><strong>m&eacute;thode</strong></em><code><span><em><strong>crier()</strong></em></span></code> comme une m&eacute;thode abstraite.</p>

In [57]:
from abc import ABCMeta, abstractmethod


class Animal(metaclass=ABCMeta):

    @abstractmethod
    def crier(self):
        pass

<p>Toute tentative de cr&eacute;er un objet de type&nbsp;<code><span><em><strong>Animal</strong></em></span></code> &eacute;chouera &agrave; l&rsquo;ex&eacute;cution:</p>

In [58]:
a = Animal()

TypeError: Can't instantiate abstract class Animal with abstract method crier

<p style="text-align: justify;"><span>La classe&nbsp;</span><code><span><em><strong>Animal</strong></em></span></code><span>&nbsp;est en quelque sorte devenu un contrat qui indique que toute classe qui en h&eacute;rite doit fournir une impl&eacute;mentation pour la m&eacute;thode&nbsp;</span><em><strong><code><span>crier()</span></code></strong></em><span>.</span></p>

In [59]:
class Chien(Animal):

    def crier(self):
        print("whouaf whouaf !")

In [60]:
c = Chien()
c.crier()

whouaf whouaf !
